src/ucs/debug/debug.c

```diff
@@ -1188,15 +1225,54 @@ void ucs_debug_init()
 void ucs_debug_cleanup(int on_error)
 {
     char *sym;
+    int signum;
+    struct sigaction *hndl;
+
+    kh_foreach_key(&ucs_signal_orig_action_map, signum,
+                   ucs_debug_disable_signal(signum));
 
-    if (ucs_global_opts.handle_errors) {
-        ucs_set_signal_handler(NULL);
-    }
-    if (ucs_global_opts.debug_signo > 0) {
-        signal(ucs_global_opts.debug_signo, SIG_DFL);
-    }
     if (!on_error) {
         kh_foreach_value(&ucs_debug_symbols_cache, sym, ucs_free(sym));
+        kh_foreach_value(&ucs_signal_orig_action_map, hndl, ucs_free(hndl));
         kh_destroy_inplace(ucs_debug_symbol, &ucs_debug_symbols_cache);
+        kh_destroy_inplace(ucs_signal_orig_action, &ucs_signal_orig_action_map);
     }
+    pthread_mutex_destroy(&ucs_kh_lock);
+}
+
+void ucs_debug_disable_signal(int signum)
+{
+    khiter_t hash_it;
+    struct sigaction *original_action, ucs_action;
+    int ret;
+
+    pthread_mutex_lock(&ucs_kh_lock);
+    hash_it = kh_get(ucs_signal_orig_action, &ucs_signal_orig_action_map,
+                     signum);
+    if (hash_it == kh_end(&ucs_signal_orig_action_map)) {
+        ucs_warn("ucs_debug_disable_signal: signal %d was not set in ucs",
+                 signum);
+        goto out;
+    }
+
+    original_action = kh_val(&ucs_signal_orig_action_map, hash_it);
+    ret = orig_sigaction(signum, original_action, &ucs_action);
+    if (ret < 0) {
+        ucs_warn("failed to set signal handler for sig %d : %m", signum);
+    }
+
+    kh_del(ucs_signal_orig_action, &ucs_signal_orig_action_map, hash_it);
+    ucs_free(original_action);
+
+out:
+    pthread_mutex_unlock(&ucs_kh_lock);
+}
+
+void ucs_debug_disable_signals()
+{
+    int signum;
+    pthread_mutex_lock(&ucs_kh_lock);
+    kh_foreach_key(&ucs_signal_orig_action_map, signum,
+                   ucs_debug_disable_signal(signum));
```

Comment:
is this the only reason this lock has to be recursive?
another option is to add an internal helper function like ucs_debug_disable_signal_nolock()..